Implement wxImageDataObject

a-wi opened the pull request:
We need a way to store images with metadata (especially image resolution) to clipboard - see [17631](https://trac.wxwidgets.org/ticket/17631). wxImage contains this extra information (unlike to wxBitmap) so we could use it for this purposes by implementing a new data object for wxImage data.
For wxGTK and wxOSX we can use their native PNG/TIFF clipboard data formats to store the image and its metadata but for wxMSW we need to implement a custom PNG data handler. It's non-standard but it is supported by several applications like MS Excel, MS PowerPoint or GIMP.

a-wi committed:
Implement wxImageDataObject

Using this object we can put an wxImage on or retrieve it from the clipboard. wxImage is stored internally as a blob with either a PNG file (wxMSW, wxGTK) or a TIFF file (wxOSX) and therefore some its metadata (like resolution) is stored on the clipboard too (what is not the case for wxBitmap stored with wxBitmapDataObject). wxImages stored with wxImageDataObject can be used by native applications.

Closes #17631.

vadz reviewed (commented):
Thanks a lot!

In addition to various nitpicks below (I hope it's not too much trouble, i.e. I think you could just commit all these suggestions), I have 2 more global questions:

1. Do we really want to duplicate exactly the same code for MSW/GTK and Mac? It looks like we could just define `IMAGE_FORMAT_BITMAP_TYPE` constant differently for them and then reuse the same code for both. I'd do this unless you foresee some divergences between them in the future.
2. I am not sure how much sense does it make to provide the new class for the platforms without support for it. I'd either just not define it at all or make it use PNG everywhere, because why not?

Last but maybe not least, I think it would be nice to add a simple test to the unit test suite putting an image on clipboard and retrieving it from it to check that the round trip works as expected.

Thanks again!

a-wi committed:
Provide support for storing PNG files on the clipboard (wxMSW)

PNG is a custom clipboard format but images stored in this format are supported by e.g. MS Excel, MS PowerPoint, GIMP (including image transparency).

See #17631.

a-wi committed:
Demonstrate storing wxImage to clipboard in image sample

a-wi committed:
Add test of putting wxImage on and retrieving it from clipboard

a-wi commented:
>     2\. I'd either just not define it at all or make it use PNG everywhere, because why not?
Indeed, PNG format could be default format for other ports than 3 major ones. I've updated the code.

vadz commented:
Thanks again!
